### PR TITLE
User Registration with OTP Email Verification and JWT Authentication

### PR DESCRIPTION
**Pull request by omkarRanu3625**
This PR adds functionality to handle **user registration**, **OTP** email verification, and **JWT** token generation for **authentication**. The following major features and changes are included in this update:
---

### 1. User Registration:
- Added a registration route (**POST /auth/register**) where users can sign up with their name, username, email, password, and phone number.
- Passwords are hashed using `bcrypt` for secure storage.
- A new user record is created in the **MongoDB** database upon successful registration.

### 2. User Login:
Added a login route (**/auth/login**) where user can login with their email and password.

### 3. Password Validation:
Passwords must meet specific security criteria, ensuring they include at least:
- One `uppercase letter`.
- One `lowercase` letter`.
- One `digit`.
- One `special character`.
- Length between `8 to 20` characters.

### 4. OTP Email Verification:

- Added two routes one is `POST /auth/verify-otp` and another is `POST /auth/resent-otp`.
- After registration, a One-Time Password (OTP) is generated and sent to the user's email for account verification.
- OTP expires in **10 minutes** and is temporarily stored (in memory, database, or Redis).
- OTP is sent via Gmail **SMTP** using the `nodemailer` package.
- Added functionality for **resending OTP** in case the user doesn't receive it or the OTP expires.

### 5. JWT Token Generation:

- JWT (JSON Web Token) is used for authentication and is generated upon successful login.
- JWT token contains user details and expires in 3 days (configurable via environment variables).
- Tokens are verified using `jsonwebtoken` to authenticate users during login.

---

## New Files:
1. `routes/auth.js`:
- Contains routes for user registration (`POST /auth/register`), login (`POST /auth/login`), verify otp (`POST /auth/verify-otp`) and resending otp (`POST /auth/resent-otp`).
2. `controllers/auth/register.js`:
- Implements the user registration logic, including password hashing, OTP generation, and email verification.
- Validates the user's password against the defined security criteria.
3. `controllers/auth/login.js`:
- Implements login functionality to authenticate users by email and password.
- Upon successful login, a JWT token is generated and returned.
4. `config/configJWT.js`:
- Contains functions to generate and verify JWT tokens for user authentication using `jsonwebtoken`.
5. `config/configSMTP.js`:
- Implements the email sending functionality using `nodemailer` to send OTPs to users for verification.

---

## Changes file:
1. `models/User.js`:
- Remove `phoneNumber`  schema attribute `minLength` and `maxLength`.
2. `index.js`:
- Added Authentication routes ` app.use('/api/auth', authRoutes)`
3. `README.md`:
- Added OTP Verification and resending route.
- Add Authentication details.

---

## .env Configuration:
Add the following environment variables to your .env file to configure the application properly and access data using `dotenv` module:
 ```
MONGO_URI= 
JWT_SECRET=
SMTP_EMAIL=
SMTP_PASSWORD=
PORT=
```
---

### CODE REVIEW

File: controllers/auth/login.js [hunk 1]
```diff
@@ -1 +1,36 @@
...
+
+  try {
+    // Check if the user exists
+    const user = await User.findOne({ email });
+    if (!user) {
+      return res.status(400).json({Status:0, Message: 'Invalid credentials' });
+    }
+
+    // Check if the user is verified
+    if (!user.isVerified) {
```
Comment (SHIVAM-KUMAR-59) [inline controllers/auth/login.js:17]: Initially every user will be not verified so please remove this check as the verification of user's email is only required while posting the jobs and not while registering or logging in

File: controllers/auth/register.js [hunk 2]
```diff
@@ -1 +1,130 @@
...
+        Message: 'User registered successfully. Please check your email for OTP verification',
+      });
+    } catch (error) {
+      return res.status(500).json({ Status: 0, Message: 'Something went wrong', error });
+    }
+  };
+  
+
+// Function to handle OTP resending
+const resendOTP = async ( req, res) => {
```
Comment (SHIVAM-KUMAR-59) [inline controllers/auth/register.js:68]: Put the resendOTP and verifyOTP functions in the utils folder, inside the utils folder make another folder for auth and inside that you can make 2 files, one for each function and then refactor here accordingly.

File: package.json [hunk 3]
```diff
@@ -19,9 +19,13 @@
   "description": "",
   "dependencies": {
     "bcrypt": "^5.1.1",
+    "bcryptjs": "^2.4.3",
     "cors": "^2.8.5",
+    "crypto": "^1.0.1",
     "dotenv": "^16.4.7",
     "express": "^4.21.2",
-    "mongoose": "^8.9.2"
```
Comment (SHIVAM-KUMAR-59) [inline package.json:25]: Have you deleted mongoose package? If yes then kindly download it again cause a lot of us are using that.

File: models/User.js [hunk 4]
```diff
@@ -20,8 +20,6 @@ const User = new mongoose.Schema(
     password: {
       type: String,
       required: true,
-      minLength: 8,
```
Comment (SHIVAM-KUMAR-59) [inline models/User.js:23]: Why did you delete this? Kindly do not make changes in the models unless very urgent.

File: index.js [hunk 5]
```diff
@@ -15,6 +16,9 @@ app.get('/', (req, res) => {
   res.status(200).send('Hello World!')
 })
 
+//Authentication Routes
```
Comment (SHIVAM-KUMAR-59) [inline index.js:19]: All the routes are to be configured in routes/api.js and then exported as one single route.

**Review by SHIVAM-KUMAR-59 — CHANGES_REQUESTED** [review]

File: controllers/auth/login.js [hunk 6]
```diff
@@ -1 +1,36 @@
 // Login Controller
+const bcrypt = require('bcryptjs');
+const User = require('../../models/User');
+const { generateToken } = require('../../config/configJWT');
+
+const login = async (req, res) => {
```
Comment (SHIVAM-KUMAR-59) [inline controllers/auth/login.js:6]: Store the token in the cookies.

File: controllers/auth/register.js [hunk 7]
```diff
@@ -1 +1,130 @@
...
+const bcrypt = require('bcryptjs');
+const User = require('../../models/User');
+const { generateToken } = require('../../config/configJWT');
+const { sendOTP } = require('../../config/configSMTP');
+const crypto = require('crypto');
+
+// Temporary storage for OTP (This can be replaced with Redis or DB for production)
+let otpStorage = {};
+
+const register = async (req, res) => {
```
Comment (SHIVAM-KUMAR-59) [inline controllers/auth/register.js:10]: Store the token in the cookie.

File: routes/auth.js [hunk 8]
```diff
@@ -1 +1,13 @@
 // Authentication routes go here
+
```
Comment (SHIVAM-KUMAR-59) [inline routes/auth.js:2]: Good job exporting these routes, import these routes in the api.js file and you may export that in the main application.

File: controllers/auth/login.js (hunk 1 above)
Comment (omkarRanu3625) [reply controllers/auth/login.js]: Done

File: controllers/auth/register.js (hunk 2 above)
Comment (omkarRanu3625) [reply controllers/auth/register.js]: Done

**Review by SHIVAM-KUMAR-59 — APPROVED** [review]

File: models/User.js (hunk 4 above)
Comment (omkarRanu3625) [reply models/User.js:23]: done

File: index.js (hunk 5 above)
Comment (omkarRanu3625) [reply index.js]: done

File: controllers/auth/login.js (hunk 6 above)
Comment (omkarRanu3625) [reply controllers/auth/login.js:6]: done
